Pattern Docs: Avoiding empty state (#1911)
hashicorp/design-system: website/docs/patterns/filter-patterns/partials/guidelines/guidelines.md

![Empty state in applied filters](/assets/patterns/filter-patterns/applied-filters-empty-state.png =450x*)

### Avoiding an empty state

> Because filtering is often reductive in nature and driven by a user action, it might not always be possible to avoid an empty state within a data set.

Reviewer: ```suggestion
As filtering is reductive in nature and driven by a user action, it’s not always possible to avoid an empty state.
```

Reviewer: I'm curious as to why you would want to "avoid" an empty state when filtering, though. If nothing is found for the filter you've set, then an empty state is valid.

> If technically possible, use the `count` property within the Dropdown ListItem to more explicitly call attention to filters that will return zero results.

Reviewer: ```suggestion
If technically feasible, use the `count` property within the Dropdown ListItem to call attention to filters that will return zero results.
```

> In this example, `status = pending` does not correspond with any results.

Reviewer: ```suggestion
In this example, "Status > Pending" does not return any results.
```

![Count property in ListItem](/assets/patterns/filter-patterns/avoid-empty-state-list-item-count.png)

### Reverting filters

> If the filters applied to a data set _do_ result in an empty state or no results being returned, provide the user with an escape route to revert clear or all filters at once or individually.

Reviewer: ```suggestion
If applying filters to a data set yields no results, offer users an option to clear all or specific filters.
```

> This can be done by adding a "Clear all" action to the [Application State](/components/application-state) that clears all applied filters at once or by including dismissible [Tags](/components/tag) for each applied filter that can be cleared by the user individually.

Reviewer: ```suggestion
This can be done by adding a "Clear all" action to the [Application State](/components/application-state) or by including dismissible [Tags](/components/tag) for each applied filter.
```

![Empty state in applied filters](/assets/patterns/filter-patterns/applied-filters-empty-state.png =450x*)

> ### Avoiding empty state

Reviewer 2: [nit]

```suggestion
### Avoiding an empty state
```

Because filtering is often reductive in nature and driven by a user action, it might not always be possible to avoid an empty state within a data set. This can be helped by leveraging specific component variants and properties, and by taking steps to prevent mistakes or errors that would result in an empty state prior to them occurring.

#### ListItem count

> Use the `count` property within the Dropdown ListItem to more clearly explicitly call attention to filters that will return zero results.

Reviewer 3: Reasons why this is not always feasible:
- there can be very strong limitations in doing this (costly operation on the backend or frontend)
- the number could be very high (hundreds of thousands?) so it would not fit or convey no information per se

```suggestion
If technically possible, use the `count` property within the Dropdown ListItem to more clearly explicitly call attention to filters that will return zero results.
```

Reviewer 2: I'd also consider dropping "clearly" or "explicitly". I don't think both are needed.

Author: Ah, good point on the technical side.

In this example, `status = pending` does not correspond with any results.

![Count property in ListItem](/assets/patterns/filter-patterns/avoid-empty-state-list-item-count.png)

> #### Approximate matching

Reviewer 3: +1 on this suggestion

> #### Approximate matching

Author: I've removed this for now

> To help resolve basic spelling mistakes, typos, and broaden the relevancy of a search term being filtered upon, consider using _approximate_ matching or "fuzzy search" when filtering by a string value.

Reviewer 2: Is there a good resource you could link to for folks who may not be as familiar with "fuzzy search"?

This can help reduce the burden on the user to accurately enter long and complex strings that may consist of seemingly random sequences of characters which are common characteristics of string values within a data set.

!!! Info

More details around the specifics of fuzzy search are beyond the scope of these guidelines, but are included to highlight a creative and user-centric search experience when filtering.

> !!!

Reviewer 2: [suggestion] I worry that "beyond the scope of these guidelines" will result in people asking for us to include it. Maybe instead, this could be reframed as something like: 

We don't offer or support fuzzy search, but we've included this concept to highlight a creative and user-centric search solution when filtering that may help the user avoid an empty state.

### Reverting filters

> If the filters applied to a data set _do_ result in an empty state or no results being returned, provide the user with an escape route to revert or all filters at once or individually.

Reviewer 2: I think there's a word missing after "revert"

> This can be done by highlighting a "Clear all" function in the [Application State](/components/application-state) to bulk clear applied filters, or individually by using a dismissible [Tag](/components/tag).

Reviewer 2: [nit] This read a bit odd to me, so I tried reframing a little.

```suggestion
This can be done by adding a "Clear all" action to the [Application State](/components/application-state) that clears all applied filters at once or by including dismissible [Tags](/components/tag) for each applied filter that can be cleared by the user individually.
```
